### DISCUSSION TITLE
Can't Connect to Redis DB - No such file/directory

### DISCUSSION THREAD

**Discussion by chisss07** [searxng/searxng#1591] [Q&A] [answered]
Hey everyone, 
I'm trying to install this on Ubuntu 22.04LTS, went through the whole step by step config. At the very end when I to go python searx/webapp.py I get the following error: 

2022-07-30 14:06:33,842 ERROR:searx.shared.redis: [searxng (999)] can't connect redis DB ...
2022-07-30 14:06:33,842 ERROR:searx.shared.redis:   Error 2 connecting to unix socket: /usr/local/searxng-redis/run/redis.sock. No such file or directory.
2022-07-30 14:06:33,842 ERROR:searx.plugins.limiter: init limiter DB failed!!!
 * Serving Flask app 'webapp' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off

I don't see the searxng-redis folder anywhere so what step did I miss where it creates that?

**Comment by ghost** [comment]
Same for me, even setting a redis URL with 6379/0 won't work.

**Reply by return42** [reply]
> `2022-07-30 14:06:33,842 ERROR:searx.shared.redis: Error 2 connecting to unix socket: /usr/local/searxng-redis/run/redis.sock. No such file or directory.`

...

> Same for me, even setting a redis URL with 6379/0 won't work.

The default connector goes to a socket:

https://github.com/searxng/searxng/blob/1fbb514a4ead209c95b4ddca0430f754a4c11554/searx/settings.yml#L79-L81

You need to set the redis URL in your `/etc/searxng/settings.yml`

**Reply by ghost** [reply]
Ha I already set it with url: redis://redis:6379/0 but the error is still there

**Reply by return42** [reply]
@JAAnon then your error message is slightly different.  Most often the searxng account from the WEB application does not have access rights to connect to the redis DB / did you checked access rights?

**Reply by ghost** [reply]
> @JAAnon then your error message is slightly different. Most often the searxng account from the WEB application does not have access rights to connect to the redis DB / did you checked access rights?

Hmm if I set limiter to false the app works well, the problem is just with redis so yes, it could be a permissions issue. How could I check all the rights?

**Reply by return42** [reply]
> How could I check all the rights?

Have a look at the redis installation instructions you followed to install redis.

**Reply by ghost** [reply]
Hey now I have a doubt..do i have to install the redis server? Because I didn't understand if only the client is available with SearxNG out of the box, maybe this is the cause of my issue?

**Reply by return42** [reply]
yes

**Comment by chisss07** [comment]
My issue is that it seems that the redis DB is nowhere to be found. That path that it’s looking for doesn’t exist and I can’t find where this is. How do I try and reinstall it?


On Jul 31, 2022, at 8:39 AM, JAAnon ***@***.***> wrote:

﻿

Ha I already set it with url: redis://redis:6379/0 but the error is still there

**Reply by return42** [reply]
If you do not have a redis DB you need to install a redis DB / If you do not have any preference you can install it from the sources https://docs.searxng.org/admin/engines/settings.html#redis

**Reply by chisss07** [reply]
You know what I lied, the folder is there, the redis.sock file is missing. How do I recreate it/get it?

**Reply by return42** [reply]
>  the redis.sock file is missing. How do I recreate it/get it?

? .. the communication socket is created by a running redis services thats listens on this socket.

**Reply by chisss07** [reply]
Yeah something is still weird, I'm going to start from scratch again and see what happens. I have a feeling I'm missing something from one of the steps. When I got to do the test at the end of the step by step I still get that the file is missing. I'll report back.

**Reply by return42** [reply]
I recommend to uninstall searxng and if you have morty and filtron.

Then give the installation a new try using our new scripts https://docs.searxng.org/admin/installation-scripts.html

**Reply by chisss07** [reply]
It seems to have worked after the second time. Thank you, not sure what was missing.

**Comment by return42** [comment, marked as answer]
> It seems to have worked after the second time. Thank you, not sure what was missing.

I assume in your first try the redis DB was not installed / thanks for your feedback on this.

**Comment by ralphrmartin** [comment]
Actually, the step-by-step instructions do not mention installing redis. Can details of how to do this please be added to the step-by-step instructions?

**Reply by return42** [reply]
Depends on your OS & distro .. or you have already one in your intranet you want to use.

The SearXNG scripts install from sources and use a socket communication https://docs.searxng.org/admin/engines/settings.html#redis

**Reply by ralphrmartin** [reply]
Well, there are explicit versions of the step-by-step instructions for several OSes, and in some sense, these instructions are incomplete if they do not also tell you how to install / setup redis; the instructions do say how to install other dependencies, and without installing redis (or having an existing one) you wont end up with a working searxng. To me "step-by-step" implies that **all** necessary steps are given, starting from scratch.

**Reply by return42** [reply]
> without installing redis (or having an existing one) you wont end up with a working searxng. 

That is not correct; redis is optional (may be in the future it becomes mandatory) .. to get in use of the limiter redis is mandatory.

A redis installation is well documented by the redis team: https://redis.io/docs/getting-started/installation/

**Reply by return42** [reply]
@ralphrmartin we discussed it internally and came to the conclusion that you are right (redis becomes more and more mandatory) :+1: .. I opened https://github.com/searxng/searxng/issues/1725

**Reply by ralphrmartin** [reply]
Thank you! 

In fact, I tried to follow the step by step instructions on Raspberry Pi OS and did not get a working result, **_maybe_** because there was no redis, so I will be happy to try again once the instructions are updated. (I followed them some time ago, before redis was used, and all worked well on Raspberry Pi OS.)

**Reply by return42** [reply]
TBH we do not test Pi OS .. but AFAIK its based on debian .. and the script and the step-by-step tutorial are tested on debian.

**Reply by ralphrmartin** [reply]
Yes, I followed the debian instructions sometime in the past for Raspberry Pi OS, and they worked fine, but since the change to using redis, they seemed incomplete. The script doesn't seem to work as RPi OS is not recognised; perhaps if it were recognised as a "variant of debian" it would work.

**Reply by return42** [reply]
> The script doesn't seem to work as RPi OS is not recognised; perhaps if it were recognised as a "variant of debian" it would work.

Thats true, user often asked for .. I really would like to support "variants" by the script .. but I can't (automated) test all variants.  I am using Linux containers to test installation procedures and I test on ..

https://github.com/searxng/searxng/blob/13ef9cc12503d8725842da6fe7c5628762b1542c/utils/lxc-searxng.env#L18-L26

Even if the variants of a distro are similar, when it goes into the details, e.g. when we look at the setup of a uWSGI the distros have their peculiarities how they build a package ([I wrote an article about.](https://docs.searxng.org/admin/installation-uwsgi.html))

Even if the probability is very high that Pi OS (a variant) builds exactly like debian (the origin), I can't blindly trust on it.

I am in a dilemma, on the one hand I would like to support more variants, especially since not much would be required. On the other hand I have no possibility to test all the variants.
